nativelib/src/main/resources/gc/commix/Sweeper.c

```diff
@@ -0,0 +1,482 @@
+U[FF]|[CC]U|U[F]U
+Batch 2 is done, thread 2 tries to coalesce. It can coalesce the remaining 2 batches.
+U[FFFF]U|U[F]U
+The free superblock of size 4 gets returned to BlockAllocator.
+U[FFFF]U|U[F]U
+U[FFFF]UU[F]U
+All is coalesced! Sweeping is done.
+
+Note that besides coalescing `Sweeper_LazyCoalesce` also finishes the sweeping of superblocks in some cases.
+See also `block_superblock_start_me` and `LargeAllocator_Sweep`.
+*/
+
+INLINE
+Object *Sweeper_LazySweep(Heap *heap, uint32_t size) {
+    Object *object = NULL;
+#ifdef ENABLE_GC_STATS
+    uint64_t start_ns, end_ns;
+    Stats *stats = heap->stats;
+    if (stats != NULL) {
+        start_ns = scalanative_nano_time();
+    }
+#endif
+    // mark as active
+    heap->lazySweep.lastActivity = BlockRange_Pack(1, heap->sweep.cursor);
+    while (object == NULL && heap->sweep.cursor < heap->sweep.limit) {
+        Sweeper_Sweep(heap, heap->stats, &heap->lazySweep.cursorDone,
+                      LAZY_SWEEP_MIN_BATCH);
+        object = (Object *)Allocator_Alloc(&allocator, size);
+    }
+    // mark as inactive
+    heap->lazySweep.lastActivity = BlockRange_Pack(0, heap->sweep.cursor);
+    while (object == NULL && !Sweeper_IsSweepDone(heap)) {
+        object = (Object *)Allocator_Alloc(&allocator, size);
+        if (object == NULL) {
+            sched_yield();
+        }
+    }
+#ifdef ENABLE_GC_STATS
+    if (stats != NULL) {
+        end_ns = scalanative_nano_time();
+        Stats_RecordEvent(stats, event_sweep, start_ns, end_ns);
+    }
+#endif
+    return object;
+}
+
+INLINE
+Object *Sweeper_LazySweepLarge(Heap *heap, uint32_t size) {
+    Object *object = NULL;
+#ifdef DEBUG_PRINT
+    uint32_t increment =
+        (uint32_t)MathUtils_DivAndRoundUp(size, BLOCK_TOTAL_SIZE);
+    printf("Sweeper_LazySweepLarge (%" PRIu32 ") => %" PRIu32 "\n", size,
+           increment);
+    fflush(stdout);
+#endif
+    // lazy sweep will happen
+#ifdef ENABLE_GC_STATS
+    uint64_t start_ns, end_ns;
+    Stats *stats = heap->stats;
+    if (stats != NULL) {
+        start_ns = scalanative_nano_time();
+    }
+#endif
+    // mark as active
+    heap->lazySweep.lastActivity = BlockRange_Pack(1, heap->sweep.cursor);
+    while (object == NULL && heap->sweep.cursor < heap->sweep.limit) {
+        Sweeper_Sweep(heap, heap->stats, &heap->lazySweep.cursorDone,
+                      LAZY_SWEEP_MIN_BATCH);
+        object = LargeAllocator_GetBlock(&largeAllocator, size);
+    }
+    // mark as inactive
+    heap->lazySweep.lastActivity = BlockRange_Pack(0, heap->sweep.cursor);
+    while (object == NULL && !Sweeper_IsSweepDone(heap)) {
+        object = LargeAllocator_GetBlock(&largeAllocator, size);
+        if (object == NULL) {
+            sched_yield();
+        }
+    }
+#ifdef ENABLE_GC_STATS
+    if (stats != NULL) {
+        end_ns = scalanative_nano_time();
+        Stats_RecordEvent(stats, event_sweep, start_ns, end_ns);
+    }
+#endif
```

Comment:
Is there any way to reduce boilerplate of stat recording via a few method-like macros? 
```
val start_ns = Stats_Start(stats);
...
Stats_Record(stats, event_sweep, start_ns);
```
Where definitions of `Stats_Start` and `Stats_Record` do nothing (i.e. just return 0) if stat recording is off?

Reply:
Currently all the ifdefs really hurt the readability of the code.